enterprise/mqtt/src/main/java/io/crate/mqtt/operations/MqttIngestService.java

```diff
@@ -19,47 +19,102 @@
-    private static final String STMT = "INSERT INTO raw (" +
-                                       "\"client_id\", \"packet_id\", \"topic\", \"ts\", \"payload\"" +
-                                       ") VALUES (" +
-                                       "?, ?, ?, CURRENT_TIMESTAMP, ?" +
-                                       ")";
-    private static final List<DataType> DATA_TYPES = Arrays.asList(DataTypes.STRING, DataTypes.INTEGER, DataTypes.STRING, DataTypes.OBJECT);
     private static final Logger LOGGER = Loggers.getLogger(MqttIngestService.class);
+    static final String SOURCE_IDENT = "mqtt";
+    private static final Map<String, Integer> MQTT_FIELDS_ORDER = ImmutableMap.of("client_id", 0,
+        "packet_id", 1,
+        "topic", 2,
+        "ts", 3,
+        "payload", 4);
+    private static final List<DataType> DATA_TYPES = Arrays.asList(DataTypes.STRING, DataTypes.INTEGER, DataTypes.STRING, DataTypes.OBJECT);
+    private static final String STATEMENT_NAME = "ingestMqtt";
+
+    private final ExpressionAnalyzer expressionAnalyzer;
+    private final InputFactory inputFactory;
+    private final SQLOperations sqlOperations;
+    private final IngestionService ingestionService;
+    private final User crateUser;
+    private final AtomicReference<Set<Tuple<Predicate<Row>, IngestRule>>> predicateAndIngestRulesReference =
+        new AtomicReference<>(new HashSet<>());
+    private final ExpressionAnalysisContext expressionAnalysisContext;
+    private boolean isInitialized;
 
-    private final Session session;
+    public MqttIngestService(Functions functions,
+                             SQLOperations sqlOperations,
+                             UserManager userManager,
+                             IngestionService ingestionService) {
+        this.sqlOperations = sqlOperations;
+        this.inputFactory = new InputFactory(functions);
+        this.expressionAnalysisContext = new ExpressionAnalysisContext();
+        FieldProvider<Symbol> mqttSourceFieldsProvider = new FieldProvider<Symbol>() {
 
-    public MqttIngestService(SQLOperations sqlOperations) {
-        session = sqlOperations.createSession(DEFAULT_SCHEMA, null, Option.NONE, 1);
-        session.parse(STMT_NAME, STMT, DATA_TYPES);
+            @Override
+            public Symbol resolveField(QualifiedName qualifiedName, Operation operation) {
+                return resolveField(qualifiedName, null, operation);
+            }
+
+            @Override
+            public Symbol resolveField(QualifiedName qualifiedName, @Nullable List<String> path, Operation operation) {
+                return new InputColumn(MQTT_FIELDS_ORDER.get(qualifiedName.toString()));
+            }
+        };
+        this.expressionAnalyzer = new ExpressionAnalyzer(functions, null, null, mqttSourceFieldsProvider, null);
+        this.ingestionService = ingestionService;
+        this.crateUser = userManager.findUser("crate");
     }
 
+    public void initialize() {
```

Comment:
I don't think we should publish an incomplete object ( exposing "this" from within the constructor can be a source of interesting bugs )

Reply:
what about doing the registration outside as well? Would also remove the dependency on the service here

Reply:
That will indeed drop the dependency. My only concern, in that case, is that there would be no way for `MqttIngestService` to know if it was registered with the `IngestionService` or not (so we'll have to rely on proper integration tests to verify that whoever is using the ingestion implementation is registering before using). As opposed to keeping the initialize internal in which case `doInsert` can differentiate between it hadn't been registered or lack of rules. 
Truth be told having to call `initialize` every time is also not too great, but at least the implementation would fail if it wasn't called.

Reply:
Not sure if that's an issue - we'll need an integration test anyway. I'm not a fan of classes which are in a "invalid" state after construction, but that's more my personal taste than anything else. So it's your call.

In case you want to keep it:
 - Should initialize also raise if it's called multiple times? (Right now it would fail anyway due to the duplicate registration)
 - Do we need/are there tests for the initialization failure? (Eg. call `doInsert` without initialization)

Either way: 
This now relies on a initial event to get the rules - is it ensured that this happens ?

Reply:
There is no initial special event but when register happens the implementation receives the existing rules